### PR TITLE
Convert fully connected models to tensorgraph

### PR DESCRIPTION
**Pull request by peastman**
I'm posting this now to get feedback on the implementation.  The new versions are currently missing two features present in the old versions: weight initialization and dropout.  I'll add those once tensorgraph supports them.  Aside from that, they seem to work correctly in my very limited testing.  The speed is significantly improved, especially when using a small batch size.  The training score is higher than before, but the validation score is slightly lower.  Presumably that's because of the lack of dropout.

I've also added support for transformers to tensorgraph.

### CODE REVIEW

File: deepchem/models/tensorflow_models/fcnet.py [hunk 1]
```diff
@@ -19,6 +19,119 @@
 from deepchem.metrics import to_one_hot
 
 
+from deepchem.models.tensorgraph.tensor_graph import TensorGraph
+from deepchem.models.tensorgraph.layers import Feature, Label, Weights, WeightedError, Dense, Reshape, SoftMaxCrossEntropy, L2LossLayer
+
+class TensorflowMultiTaskClassifier2(TensorGraph):
```
Comment (rbharath) [inline deepchem/models/tensorflow_models/fcnet.py]: Maybe call this `TensorgraphMultiTaskClassifier` to avoid the `2`.

File: deepchem/models/tensorflow_models/fcnet.py [hunk 2]
```diff
@@ -19,6 +19,119 @@
...
+        if X_b is not None:
+          feed_dict[self.features[0]] = X_b
+        if w_b is not None and not predict:
+          feed_dict[self.task_weights[0]] = w_b
+        yield feed_dict
+
+
+
+
+class TensorflowMultiTaskRegressor2(TensorGraph):
```
Comment (rbharath) [inline deepchem/models/tensorflow_models/fcnet.py]: `TensorgraphMultiTaskRegressor`

**Comment by peastman** [comment]
My plan is that these classes will completely replace the other ones.  Once the new versions are feature complete and fully validated, there's no reason to keep the old versions around.  So "migrating" will just mean deleting the old classes and renaming these ones.

**Comment by peastman** [comment]
Now that #521 is merged I can update this to use the same weight initialization as the old models.  But first I just want to confirm: do we actually think that initialization is better than the default one in tensorgraph?

**Comment by peastman** [comment]
I'm trying to implement dropout and running into a problem.  The dropout layer is only supposed to be present during training, not evaluation.  The current models handle this based on the `training` argument passed to `build()`.  But tensorgraph doesn't seem to have anything equivalent.  In fact, as far as I can tell it only ever builds the graph once and then reuses it for both training and evaluation.

**Comment by rbharath** [comment]
I think the old initialization is likely better than the current `TensorGraph` version, but I couldn't tell for sure (perhaps testing on Tox21 would be a good test).

The dropout handling is a tricky issue. The way Keras handles this is to add a under-the-hood placeholder `training` to the computational graph that marks whether or not the network is training. The `Dropout` layer then references this placeholder to perform the correct behavior. I think we should add a similar mechanism to `TensorGraph`. Would only require small updates to `build()` and `fit()` to create and populate the `training` placeholder correctly.

**Comment by peastman** [comment]
That makes sense.  The placeholder would be set to either 1.0 (for training) or 0.0 (for evaluating).  The `keep_prob` argument to `tf.nn.dropout()` would then be set to `1.0-dropout_prob*training`?

I think it makes sense to implement this as a core feature of tensorgraph rather than rebuilding it for each model.  At the beginning of `build()` it would create the placeholder which could then be accessed as `_training_placeholder`.  The dropout layer's `_create_tensor()` method would retrieve it and set its inputs accordingly.  Sound reasonable?

**Comment by rbharath** [comment]
Yep, this design sounds reasonable to me. Handling this as a core feature in `TensorGraph` makes sense since dropout is such a fundamental primitive.

**Comment by peastman** [comment]
This requires `_create_tensor()` to have a reference to the TensorGraph it's part of.  Currently it doesn't have that.  I can see a few possibilities:

- The user explicitly provides a reference to it when creating the dropout layer (minimal change to anything else, but ugly and inconvenient to use).
- Modify the Layer API so `_create_tensor()` receives it as an argument.
- Add a method to Layer called `added_to_graph()` that gets called when it is first added to the graph.  That lets it store a reference to the graph internally.

We should think about how this interacts with #523.  That tries to make layers independently usable outside of a TensorGraph, but this particular layer only works if it's part of a TensorGraph.

**Comment by rbharath** [comment]
Checked the Keras Dropout (https://github.com/fchollet/keras/blob/83001d195c7741e16cb22ee4ecc9083b27797c27/keras/layers/core.py#L103). Looks like it just refers to a global variable for `training_phase` under the hood. Not a good solution for us since we want to be clean with our scopes.

I like the idea of adding an optional `training_phase` argument to `create_tensor`. Then `TensorGraph` can pass this in to layers (most of which will ignore the argument). This shouldn't conflict with #523, since users using the layers in TF can take on the responsibility of setting this phase if it matters.

**Comment by peastman** [comment]
If we want to take that approach, how about just making the signature `create_tensor(self, **args)`?  There's a good chance we'll find other cases in the future where a layer needs to have a special value passed in.  This way it will be forward compatible.  We can pass in whatever arguments we need, and existing layers will just ignore them.

**Comment by rbharath** [comment]
+1, I like the signature `create_tensor(self, **kwargs)`. This will give us flexibility moving forward.

**Comment by peastman** [comment]
I've been testing `sampl_tf_models.py`, which uses TensorflowMultitaskRegressor, to compare the performance of the new and old implementations.  On this example, when I initially switched to using the old parameter initialization (rather than the one built into TensorGraph), that produced terrible performance, much worse than either the old implementation or the new one with default initialization.  After debugging for a while, I realized there's a strange inconsistency in how the old classes handle this.

When you request n layers, you really get n+1, the final one being used to generate the output.  So how does it initialize the final layer?  By copying the values from the last regular layer.  You can see this here:

https://github.com/deepchem/deepchem/blob/master/deepchem/models/tensorflow_models/fcnet.py#L194-L195

Notice those lines refer to `weight_init_stddevs[i]` and `bias_init_consts[i]`.  But it's outside the loop over `i`.  I suspect this was just done by accident: someone copied over the code without noticing that.  So `i` will have the value it had at the end of the preceding loop, and it will reuse the initialization values from the last regular layer.

It turns out the bias initialization constant has a huge effect on the quality of the results, and you actually get the best results when it's different in the regular and output layers.  I ran tests setting bias_init to either 0.0 (the TensorGraph default) or 1.0 (the TensorFlowGraphModel default) in each of the layers.  Here are the resulting train and validation scores:

|Regular Bias Init|Output Bias Init|Train Score|Validation Score|
|-----------------|----------------|-----------|----------|
|0.0|0.0|0.977|0.784|
|1.0|0.0|0.840|0.710|
|0.0|1.0|0.985|0.808|
|1.0|1.0|0.882|0.771|

So overall the results are best if you set the regular bias init to 0 and the output layer bias init to 1.  And each of those seems to be an independent factor affecting the results.

**Comment by peastman** [comment]
I've checked in support for dropout and customizing initialization.  We still need to decide how to deal with the inconsistency in initialization described above.  Aside from that, there are still a few features supported by the old classes but not the new ones.  `TensorflowGraphModel` has the following constructor arguments with no equivalents in `TensorGraph`: `logdir`, `penalty`, `penalty_type`, `momentum`, `optimizer`, `pad_batches`, and `verbose`.

It looks like `penalty` and `penalty_type` are used for doing additional regularization.  The default value is 0.  Is that something we should add?  If so, does it belong in TensorGraph itself or in the individual models?  Then there are methods for customizing the optimizer.  If we want to add those, they would need to be features of TensorGraph itself.  Logging and checkpointing seem to be done a bit differently in TensorGraph, so I'm not sure if `logdir` is necessary or not.  And when would you want to use the `pad_batches` argument?

**Comment by peastman** [comment]
Assuming the tests pass, I think this can be merged.  I've created new versions of the classes, but we'll still need to do a lot of work to validate them and convert the examples over to using them.  I've tested them with a few examples and they basically seem to work, but unfortunately there are no test cases for any of these classes, so I'm not at all confident the new versions fully match the behavior of the old ones.

**Comment by rbharath** [comment]
Looks good to merge in once tests pass

**Comment by peastman** [comment]
Can tensorgraph/models/robust_multitask.py be deleted?  It looks like that's obsolete code that provided a much less complete implementation of MultiTaskClassifier using the old version of the TensorGraph API.

**Comment by rbharath** [comment]
The model on there is misnamed. It's actually implementing a bypass model. Could you rename it "BypassMultitaskClassifier" for now?

**Comment by rbharath** [comment]
We should fix the model and update it to the new API and add tests though. I'd be OK with deleting or commenting out the file for now and fixing it in a follow-up PR.

**Comment by peastman** [comment]
For the moment I'll comment it out.  It will need to be rewritten since it both uses the wrong API and works in the wrong way (MultiTaskTensorGraph is obsolete).

**Comment by peastman** [comment]
I've been going through the remaining test failures, and so far they all look like bad tests, nothing related to this PR.  They fit a model against a tiny amount of data for a small number of iterations, then check the result against a tolerance that just happens to work sometimes, but not always.  In all the ones I've checked so far, if I increase the number of epochs from 10 to 1000, they pass reliably.  I'll see if I can get everything to work.

**Comment by rbharath** [comment]
Sounds reasonable. These are all newer tests which have been failing sporadically on different PRs. If there's a way to make them robust, that would be great, but if we need to comment them out, that should be OK as well.

**Comment by peastman** [comment]
Whew!  Tests are all passing now.

**Comment by peastman** [comment]
Hold off on merging this a bit longer.  I'm adding some tests for the new classes.

**Comment by peastman** [comment]
TensorGraph doesn't implement `reload()`, so I'm trying to implement it.  In the process, though, I've found there's some ambiguity in what `save()` and `reload()` are supposed to do.  For other Model subclasses, `save()` means to write out the current values of all parameters.  That is, write a checkpoint.  `reload()` then loads those saved parameters back in.  But in TensorGraph, `save()` does something completely different: it pickles the TensorGraph object itself, so a description of all the layers that make up the model.  It then has a `load_from_dir()` method for reconstructing pickled models, but as far as I can tell, no method for loading checkpoints.  Instead, it automatically reloads the most recent checkpoint (if there is one) when you call `fit()` or `predict()`.

How should this work?

**Comment by peastman** [comment]
Ok, this is truly ready to merge now.  The issue with `save()` and `reload()` mentioned above still needs to be resolved, but that's a matter for a different PR.

**Comment by rbharath** [comment]
LGTM.

**Comment by rbharath** [comment]
@peastman Hmm, good point. The `fit()` code saves checkpoints to `self.model_dir` which are then automatically loaded as you note above. Is there a use-case where the old semantics let us do something the current semantics don't?

**Comment by peastman** [comment]
I wasn't arguing for one being better than the other, just that we should pick one or the other and be consistent with it.  `save()` ought to mean the same thing for all models.  Maybe open a separate issue where we can discuss how this ought to work?
